### PR TITLE
Add transcritical versions of exisiting subcritical heat pumps  #5

### PR DESCRIPTION
**Pull request by bilwint**
#5 
Implemented Transcritical versions

created trans models
created input parameters
created topologies
updated the models in variables.py
implemented code in hp_dashboard.py
In transcritical versions, especially with intermediate pressure or temperature states, there is a chance of encountering an error (negative mass flow rates) when the temperature out of trans_HE (T_A0) falls below a certain point (where T_A0 is directly related to the heat sink backflow through ttd_l). This happens because the state after mid-expansion will be outside the saturated liquid curve. Outside the saturated liquid curve, the value of enthalpy increases with expansion, causing the outlet temperature after expansion to increase slightly. To avoid this, I calculated the possible minimum value of T_A0 from the enthalpy of saturated liquid at intermediate pressure and gave this value to the min_value to the slider heat sink backflow.
The code seems bulky. If you have a better idea, please suggest it and I will work on it

To do later
The models works perfectly with the working fluid within the params. They also need to be tested with other working fluids. Additionally, the off-design conditions need to be checked once the generalized off-design method is completed.

### CODE REVIEW

**Comment by bilwint** [comment]
topologies not created, instead used same subcritical topology

**Comment by jfreissmann** [comment]
Thank you for the extra effort of redoing the PR. I think it is very clean now.

Regarding your remarks about the minimum backflow temperature of the heat sink:

You are right that there should be some kind of mechanism to avoid ending up outside the liquid-vapor-region after the first expansion stage. I also agree with your assesment, that your current implementation is a bit bulky.
One thing that could help readability in any case would be separating the creation of the sliders and the ``invalid_temp_diff`` check from the if-clauses for this trans temp check and just initializing ``T_min_trans_out`` (should be renamed in that case) to 0, if we don't have a transcritical heat pump. Additionally, it could make sense to move the calculation of the correct `T_min_trans_out` into a separate function, so that the front end dashboard design logic stays clear.
Besides that, I am not 100% sure if the method of calculating `T_min_trans_out` is fully correct. It isn't fully clear to me, why you have to distinguish between the simple case or the intercooler and any other one cycle heat pump. Why do they have to be calculated differently, but there is no difference between setups with and without an inter heat exchanger after the gas cooler/condenser? Maybe this is just a mistake on my part and I don't fully grasp your current logic. In that case you can clear it up a bit for me maybe.

Best regards.

**Comment by jfreissmann** [comment]
I understand what you are trying to do. I think, it could be helpful for almost any heat pump model, as even one stage compression variants could have a high enough evaporation pressure, that the liquid-vapor-region is not reached after expansion. The models with an intermediate stage are just more prone to this error. You would have to calculate this minimum temperature from different points in the cycle depending on the number of stages. And if there is an internal heat exchanger, its subcooling is an important factor as well. That could mean a lot of custom calculation depending on the respective heat pump topology.

Maybe the cleanest way to implement that could be a helper method in the respective heat pump class. Possibly there is a more general way that could fit into the dashboard as well, but I don't see it clearly right now. The other options is, that we let users run into this issue and give hints, so that they could correct it themselves, but I'm not sure if that is the best user experience. What do you think?

**Comment by bilwint** [comment]
I also think it would be better if the user knows the issue and can act accordingly. So I will look into both options: implementing a hint message in the dashboard before running the simulation, and, as you suggested, providing a message with a hint to overcome the issue after running the simulation.
I will work on it and will let you know.

**Comment by bilwint** [comment]
I recently made four commits regarding the condition checking.

What I did is, I created generalized function in Heat pump base.  Which then calls in each heat pump model. 

The conditions i used are,
The inlet temperature of the expansion valve should be greater than the saturation temperature corresponding evaporator pressure or the mid pressure
p_mid < p_crit
T_mid < T_crit_wf1


if this is okay I will clean the code and will make the pull request. or If you have suggestions I will also look into that.

**Comment by bilwint** [comment]
One thing I forgot. 

In dashboard, I used a " if " condition. Because at the time I did only for transcritical model. After doing the same in subcritical, I think we can remove the" if " clause.

File: src/heatpumps/models/HeatPumpCascade2IHXTrans.py [hunk 1]
```diff
@@ -646,4 +648,75 @@ def generate_state_diagram(self, refrig='', diagram_type='logph',
...
+    #         errors.append(f'Error: Mid Temerature should be less than the '
+    #         f'critical temeperature {round(T_crit_wf1, 2)} of lower cycle working fluid')
+    #     else:
+    #         pass
+    #     if errors:
+    #         return errors
+    #     else:
+    #         return 'Die Simulation der Wärmepumpenauslegung war erfolgreich.'
+
+    def simulation_condiion_check(self):
```
Comment (jfreissmann) [inline src/heatpumps/models/HeatPumpCascade2IHXTrans.py:701]: There seems to be a typo here: condiion -> condition

File: src/heatpumps/models/HeatPumpCascade2IHXTrans.py [hunk 2]
```diff
@@ -646,4 +648,75 @@ def generate_state_diagram(self, refrig='', diagram_type='logph',
...
+        errors.update(cycle2_err)
+        cycle1_err = self.evap_state_condition_check(
+            conn_valve_in='D2', p_evap=self.p_evap1, wf=self.wf1
+        )
+
+        errors.update(cycle1_err)
+
+        T_crit_wf1 = PSI('T_critical', self.wf1) - 273.15
+        if self.T_mid > T_crit_wf1:
+            errors.update(f'Error: Mid Temerature should be less than the '
```
Comment (jfreissmann) [inline src/heatpumps/models/HeatPumpCascade2IHXTrans.py:715]: Temerature -> temperature

File: src/heatpumps/models/HeatPumpCascade2IHXTrans.py [hunk 3]
```diff
@@ -646,4 +648,75 @@ def generate_state_diagram(self, refrig='', diagram_type='logph',
...
+                          f'critical temeperature {round(T_crit_wf1, 2)} of lower cycle working fluid')
```
Comment (jfreissmann) [inline src/heatpumps/models/HeatPumpCascade2IHXTrans.py:716]: temeperature -> temperature

File: src/heatpumps/models/HeatPumpCascadeTrans.py [hunk 4]
```diff
@@ -586,3 +588,73 @@ def generate_state_diagram(self, refrig='', diagram_type='logph',
...
+    #         errors.append(f'Error: Mid Temerature should be less than the '
+    #                       f'critical temeperature {round(T_crit_wf1, 2)} of lower cycle working fluid')
+    #     else:
+    #         pass
+    #     if errors:
+    #         return errors
+    #     else:
+    #         return 'Die Simulation der Wärmepumpenauslegung war erfolgreich.'
+
+    def simulation_condiion_check(self):
```
Comment (jfreissmann) [inline src/heatpumps/models/HeatPumpCascadeTrans.py:639]: Same typo as above.

File: src/heatpumps/models/HeatPumpEconTrans.py [hunk 5]
```diff
@@ -510,21 +512,74 @@ def get_plotting_states(self, **kwargs):
...
+    #         errors.append(f'Error: The temperature before mid expansion {round(T_mid_valve_in,2)}°C '
+    #                       f'should be greater than the saturation temperature {round(T_sat_p_mid,2)}°C '
+    #                       f'corresponding mid pressure')
+    #
+    #     if errors:
+    #         return errors
+    #     else:
+    #         return 'Die Simulation der Wärmepumpenauslegung war erfolgreich.'
+
+    def simulation_condiion_check(self):
```
Comment (jfreissmann) [inline src/heatpumps/models/HeatPumpEconTrans.py:565]: Same type as above.

File: src/heatpumps/models/HeatPumpCascade2IHXTrans.py [hunk 6]
```diff
@@ -646,4 +648,75 @@ def generate_state_diagram(self, refrig='', diagram_type='logph',
...
+        else:
+            pass
+        if errors:
+            return errors
+        else:
+            return f'Die Simulation der Wärmepumpenauslegung war erfolgreich.'
```
Comment (jfreissmann) [inline src/heatpumps/models/HeatPumpCascade2IHXTrans.py:722]: This does not have to be an f-string.

File: src/heatpumps/models/HeatPumpBase.py [hunk 7]
```diff
@@ -948,3 +948,152 @@ def validate_dir(self):
...
+            T_sat_p_mid = PSI(
+                'T', 'Q', 0, 'P', p_mid * 1e5,
+                wf) - 273.15
+        else:
+            pass
+
+        if T_mid_valve_in > T_sat_p_mid:
+            pass
+        else:
+            errors.add(f'Error: The temperature before mid expansion {round(T_mid_valve_in, 2)}°C '
```
Comment (jfreissmann) [inline src/heatpumps/models/HeatPumpBase.py:1090]: Is there a reason you are using `sets` and not `lists` or `tuples`? Also, why are you not concatinating strings with the `+` operator? Finally, you can round floats in f-strings with this notation: `f'... {T_mid_value_in:.2f}°C'`, with the number behind the dot setting the number of decimal places after the decimal point.

Comment (bilwint) [reply src/heatpumps/models/HeatPumpBase.py:1090]: I initially did it with **'list'**. At that time even if there is no error. I got error msg with with an empty list (**"[ ]"**). That means even if the simulation is Successful, gets error.

**Comment by jfreissmann** [comment]
Hey @bilwint, did you do all the changes combined in that commit?

**Comment by bilwint** [comment]
yes, I have did all the raised problems.

**Pull request merged**